doc/prebuilt/man/taler-auditor-exchange.1
.\" Man page generated from reStructuredText.
.
.TH "TALER-AUDITOR-EXCHANGE" "1" "Mar 22, 2020" "0.6pre1" "GNU Taler"
.SH NAME
taler-auditor-exchange \- add or remove exchange from auditor’s list
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
\fBtaler\-auditor\-exchange\fP [\fB\-h\fP\ |\ \fB–help\fP] [\fB\-r\fP\ |\ \fB–remove\fP]
[\fB\-m\fP\ \fIMASTERKEY\fP\ |\ \fB–exchange\-key=\fP‌\fIMASTERKEY\fP]
[\fB\-u\fP\ \fIEXCHANGE_URL\fP\ |\ \fB–auditor\-url=\fP‌\fIEXCHANGE_URL\fP]
.SH DESCRIPTION
.sp
\fBtaler\-auditor\-exchange\fP is a command line tool to be used by an
auditor to add or remove an exchange from the list of exchange’s audited
by the auditor. You must add an exchange to that list before signing
denomination keys with taler\-auditor\-sign or trying to audit it with
taler\-auditor or taler\-wire\-auditor. Afterwards the exchange will be
visible via the /exchanges API of the taler\-auditor\-httpd.
.INDENT 0.0
.TP
\fB\-m\fP \fIMASTERKEY\fP | \fB–exchange\-key=\fP‌\fIMASTERKEY\fP
Public key of the exchange in Crockford base32 encoding, for example
as generated by gnunet\-ecc \-p.
.TP
\fB\-h\fP | \fB–help\fP
Print short help on options.
.TP
\fB\-u\fP \fIEXCHANGE_URL\fP | \fB–auditor\-url=\fP‌\fIEXCHANGE_URL\fP
URL of the exchange. The exchange’s HTTP API must be available at
this address.
.TP
\fB\-r\fP | \fB–remove\fP
Instead of adding the exchange, remove it. Note that this will drop
ALL data associated with that exchange, including existing auditing
information. So use with extreme care!
.UNINDENT
.SH DIAGNOSTICS
.sp
\fBtaler\-auditor\-exchange\fP will return 0 on success, 1 on usage errors, 3 on problems interacting with the database backend, 4 if exchange entry to be added is already in the database (or already missing when used with \fB\-r\fP).
.SH SEE ALSO
.sp
gnunet\-ecc(1), taler\-auditor\-sign(1), taler.conf(5)
.SH BUGS
.sp
We should optionally verify the correctness of this exchange’s base URL
and that it matches the master public key (note that the exchange may
still be offline, so it should be possible to bypass such a verfication
step). Furthermore, if we do verification, as a (less secure)
convenience option, we should make \fB\-\fP m optional and obtain it from
the base URL.
.sp
Report bugs by using \fI\%https://gnunet.org/bugs/\fP or by sending electronic
mail to <\fI\%taler@gnu.org\fP>.
.SH AUTHOR
GNU Taler contributors
.SH COPYRIGHT
2014, 2015, 2016 Florian Dold, Benedikt Muller, Sree Harsha Totakura, Christian Grothoff, Marcello Stanisci (GPLv3+ or GFDL 1.3+)
.\" Generated by docutils manpage writer.
.


ChangeLog
Sat 14 Nov 2020 05:47:30 PM CET
    Modify taler-exchange-transfer to continue even after a
    wire transfer failed due to the bank refusing it because
    the target account does not exist. Changed the database
    to track such failures in the respective table.
    Opens new issue #6647.  -CG

Tue 10 Nov 2020 01:03:22 PM CET
    Updates to error codes and HTTP status codes for improved
    consistency. Fixed spelling issues. Ensure main() returns
    0 when called with '-h' or '--help'.
    Releasing GNU Taler 0.8.1. -CG

Sat 03 Oct 2020 03:59:45 PM CEST
    Various minor fixes, code cleanup, updates to more recent
    GNUnet APIs, new error codes, and timetravel test support.
    Releasing GNU Taler 0.8.0. -CG

Sun 29 Mar 2020 08:53:46 PM CEST
    Changed protocol to be more RESTful. Expanded auditor tests.
    Completed transition to new wire gateway API for bank interaction.
    Releasing GNU Taler 0.7.0. -CG

Tue 24 Dec 2019 11:09:14 PM CET
    Releasing GNU Taler 0.6.0. -CG

Sat 17 Aug 2019 10:03:38 PM CEST
    Remove "currency" field from exchange database, as we only
    support one currency per exchange anyway. -CG

Wed 26 Jun 2019 03:31:52 PM CEST
    Adding link signatures to prevent exchange from tracking
    users using coins falsely believed to have been recovered via /link,
    bumping protocol version to 4. -CG

Sat 08 Jun 2019 07:54:33 PM CEST
    Change payto://sepa/ to payto://iban/ as per current spec. -CG

Fri 03 May 2019 05:36:10 PM CEST
    Add support for compressing bodies during HTTP uploads to
    exchange and auditor. -CG

Thu 02 May 2019 09:15:37 PM CEST
    Pass hash of denomination public key in operations from
    clients to exchange instead of the (larger) full public key.
    Breaks protocol compatibility, bumping protocol to v3. -CG

Mon Mar 11 03:24:07 CET 2019
	Completed implementation of #5536 (delete private keys once we
	no longer need them). -CG

Sat Mar  2 19:09:43 CET 2019
	Changing denomination key revocation file format and moving them
	to their own directory (preparations for #5536 resolution). -CG

Sat Feb 23 14:47:12 CET 2019
	Actually disable 100 CONTINUE. -CG

Sun Aug 19 15:15:48 CEST 2018
	Increase various limits and rework transaction scopes to
	improve scalability. -CG

Tue Apr  3 23:29:06 CEST 2018
	Releasing Taler v0.5.0

Thu Jan  4 11:55:41 CET 2018
	Fix issue #5234 (aggregator ignoring refunds).
	Misc. minor fixes to the auditor. -CG

Mon Jan  1 23:15:37 CET 2018
	Add TALER_EXCHANGE_refund2() API call to libtalerexchange. -CG

Thu Dec 14 15:32:50 CET 2017
	Replaced taler-exchange-reservemod tool with new taler-bank-transfer
	tool (#5195).  Removed /admin/add/incoming API, replaced by new
	taler-exchange-wirewatch tooling. (#5077).

Sat Dec  9 15:21:50 CET 2071
	Implement refresh protocol optimization, bumping protocol version
	to v2 (#5178). -CG

Thu Nov  2 17:39:40 CET 2017
	Limit amount values to 2^53 as we always wanted (#5167). -CG

Thu Nov  2 16:50:46 CET 2017
	Fixing serious bug in reserve balance calculations, as
	fees were not always properly included in the summation. -CG

Tue Oct 31 13:43:55 CET 2017
	Extended crypto API to provide planchet generation functions
	explicitly (to be used on withdraw, refresh and tipping).
	Renamed a few API calls to better reflect their functions.
	Changed a few function signatures to use new structs. -CG

Wed Oct 18 15:20:29 CEST 2017
	Releasing Taler v0.4.0. -CG

Tue Oct 17 14:32:46 CEST 2017
	Fix building of /keys response to include full auditor
	signature set instead of partial map. -CG

Mon Oct 16 12:10:35 CEST 2017
	Realize JSON-based report formats from auditor (#4963). -CG

Sun Sep 17 16:46:13 CEST 2017
	Implement /keys cherry picking (#4840). -CG

Wed Sep 13 14:06:44 CEST 2017
	Add argument to TALER_EXCHANGE_check_keys_current()
	to force download even if /keys are still considered
	current. -CG

Fri Jul 14 17:38:54 CEST 2017
	Fix #4751 by not allowing denomination deletion (GC)
	to cascade into reserve_out table (and tolerating such
	deletion failures). This way, denominations will be kept
	around until the last reserve that references them (or
	known_coins derived from them) is closed. Naturally, in
	most case we expect reserves to be closed much faster
	than DKs, so in practice this should rarely apply. -CG

Sun Jun 11 17:03:56 CEST 2017
	Finish implementation and testing of automated refunding
	of expired reserves (#4956). -CG

Tue Jun  6 13:53:34 CEST 2017
	Releasing taler-exchange 0.3.0. -CG

Mon Apr 17 01:29:07 CEST 2017
	Add support for HTTP body compression (#4982). -CG

Mon Mar 20 04:37:46 CET 2017
	Implemented first working version of taler-auditor. -CG

Sat Mar 18 03:44:38 CET 2017
	Add 'wire_out' table to exchange DB to track outgoing
	wire transfers. -CG

Fri Nov 18 18:53:30 CET 2016
	Improved error reporting for bogus wire specifications.
	Releasing taler-exchange 0.2.0. -CG

Thu Nov 17 16:37:22 CET 2016
	Added missing idempotency checks for /refresh/reveal. -CG

Thu Nov 17 11:37:56 CET 2016
	Fixed a few cases of missing database rollbacks, causing the
	exchange to be stuck. -CG

Mon Jun 20 17:40:49 CEST 2016
 	Releasing taler-exchange 0.0.2. -CG

Mon Jun 13 16:35:09 CEST 2016
	Splitting HTTP service to run on different listen
	sockets for /admin and /test-APIs and the normal public
	APIs (#4569). -CG

Thu Jun  9 19:17:02 CEST 2016
	Rename /deposit/wtid to /track/transaction and
	/wire/deposits to /track/transfer. -CG

Wed Jun  8 11:32:41 CEST 2016
	Export libtalerfakebank API so we can use it in testcases for the
	merchant as well. -CG

Tue Jun  7 15:13:46 CEST 2016
        Adding public key of the exchange that was used to sign replies
        to applicable callbacks of libtalerexchange. (This will eventually
        be needed by the merchant's backend.) -CG

Mon Jun  6 23:59:05 CEST 2016
	Minor code cleanup, fixing FTBFS when run with just 'make'
	or when header installed to a location not included.
	Releasing taler-exchange 0.0.1. -CG


src/auditor/taler-auditor-exchange.c
/*
  This file is part of TALER
  Copyright (C) 2014, 2015, 2018, 2019 Taler Systems SA

  TALER is free software; you can redistribute it and/or modify it under the
  terms of the GNU General Public License as published by the Free Software
  Foundation; either version 3, or (at your option) any later version.

  TALER is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
  A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

  You should have received a copy of the GNU General Public License along with
  TALER; see the file COPYING.  If not, see <http://www.gnu.org/licenses/>
*/
/**
 * @file taler-auditor-exchange.c
 * @brief Tool used by the auditor to add or remove the exchange's master key
 *        to its database.
 * @author Christian Grothoff
 */
#include <platform.h>
#include "taler_exchangedb_lib.h"
#include "taler_auditordb_lib.h"


/**
 * URL of the exchange.
 */
static char *exchange_url;

/**
 * Master public key of the exchange.
 */
static struct TALER_MasterPublicKeyP master_public_key;

/**
 * Our configuration.
 */
static struct GNUNET_CONFIGURATION_Handle *cfg;

/**
 * Handle to access the auditor's database.
 */
static struct TALER_AUDITORDB_Plugin *adb;

/**
 * -r option given.
 */
static int remove_flag;


/**
 * The main function of the taler-auditor-exchange tool.  This tool is used
 * to add (or remove) an exchange's master key and base URL to the auditor's
 * database.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, non-zero on error
 */
int
main (int argc,
      char *const *argv)
{
  char *cfgfile = NULL;
  const struct GNUNET_GETOPT_CommandLineOption options[] = {
    GNUNET_GETOPT_option_cfgfile (&cfgfile),
    GNUNET_GETOPT_option_help (
      "Add or remove exchange to list of audited exchanges"),
    GNUNET_GETOPT_option_mandatory
      (GNUNET_GETOPT_option_base32_auto ('m',
                                         "exchange-key",
                                         "KEY",
                                         "public key of the exchange (Crockford base32 encoded)",
                                         &master_public_key)),
    GNUNET_GETOPT_option_string ('u',
                                 "exchange-url",
                                 "URL",
                                 "base URL of the exchange",
                                 &exchange_url),
    GNUNET_GETOPT_option_flag ('r',
                               "remove",
                               "remove the exchange's key (default is to add)",
                               &remove_flag),
    GNUNET_GETOPT_option_version (VERSION "-" VCS_VERSION),
    GNUNET_GETOPT_OPTION_END
  };

  TALER_gcrypt_init (); /* must trigger initialization manually at this point! */
  GNUNET_assert (GNUNET_OK ==
                 GNUNET_log_setup ("taler-auditor-exchange",
                                   "WARNING",
                                   NULL));
  {
    int ret;

    ret = GNUNET_GETOPT_run ("taler-auditor-exchange",
                             options,
                             argc, argv);
    if (GNUNET_NO == ret)
      return 0;
    if (GNUNET_SYSERR == ret)
      return 1;
  }
  if (NULL == cfgfile)
    cfgfile = GNUNET_strdup (GNUNET_OS_project_data_get ()->user_config_file);
  cfg = GNUNET_CONFIGURATION_create ();

  GNUNET_log (GNUNET_ERROR_TYPE_INFO,
              "Loading config file: %s\n",
              cfgfile);

  if (GNUNET_SYSERR ==
      GNUNET_CONFIGURATION_load (cfg,
                                 cfgfile))
  {
    GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
                "Malformed configuration file `%s', exit ...\n",
                cfgfile);
    GNUNET_free (cfgfile);
    return 1;
  }
  GNUNET_free (cfgfile);

  if (! remove_flag)
  {
    if (NULL == exchange_url)
    {
      fprintf (stderr,
               _ ("Missing either `%s' or `%s'.\n"),
               "-u URL",
               "--remove");
      return 1;
    }
    if ( (0 == strlen (exchange_url)) ||
         ( (0 != strncasecmp ("http://",
                              exchange_url,
                              strlen ("http://"))) &&
           (0 != strncasecmp ("https://",
                              exchange_url,
                              strlen ("https://"))) ) ||
         ('/' != exchange_url[strlen (exchange_url) - 1]) )
    {
      fprintf (stderr,
               "Exchange URL must begin with `http://` or `https://` and end with `/'\n");
      return 1;
    }
  }


  if (NULL ==
      (adb = TALER_AUDITORDB_plugin_load (cfg)))
  {
    fprintf (stderr,
             "Failed to initialize auditor database plugin.\n");
    return 3;
  }

  /* Create required tables */
  if (GNUNET_OK !=
      adb->create_tables (adb->cls))
  {
    GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
                "Failed to create tables in auditor's database\n");
    TALER_AUDITORDB_plugin_unload (adb);
    return 3;
  }

  /* Update DB */
  {
    enum GNUNET_DB_QueryStatus qs;
    struct TALER_AUDITORDB_Session *session;

    session = adb->get_session (adb->cls);
    if (NULL == session)
    {
      GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
                  "Failed to initialize database session\n");
      TALER_AUDITORDB_plugin_unload (adb);
      return 3;
    }

    if (remove_flag)
    {
      qs = adb->delete_exchange (adb->cls,
                                 session,
                                 &master_public_key);
    }
    else
    {
      qs = adb->insert_exchange (adb->cls,
                                 session,
                                 &master_public_key,
                                 exchange_url);
    }
    if (0 > qs)
    {
      GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
                  "Failed to update auditor database (status code: %d)\n",
                  qs);
      TALER_AUDITORDB_plugin_unload (adb);
      return 3;
    }
    if (0 == qs)
    {
      GNUNET_log (
        GNUNET_ERROR_TYPE_WARNING,
        (remove_flag)
        ? "Could not remove exchange from database: entry already absent\n"
        : "Could not add exchange to database: entry already exists\n");
      TALER_AUDITORDB_plugin_unload (adb);
      return 4;
    }
  }
  TALER_AUDITORDB_plugin_unload (adb);
  return 0;
}


/* end of taler-auditor-exchange.c */
